File: UnitySample/Assets/Samples/FacebookLogin/FacebookLoginTest.cs
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0

using System.Collections;
using System.Collections.Generic;
using UnityEngine;
using UnityEngine.Networking;
using UnityEngine.UI;

public class FacebookLoginTest : MonoBehaviour
{
    public string loginEndpointUrl;
    public Text logOutput;

    // Start is called before the first frame update
    void Start()
    {
        // Set the login endpoint
        Debug.Log("Setting login endpoint");
        AWSGameSDKClient.Instance.Init(loginEndpointUrl, this.OnLoginOrRefreshError);

        // Get a new guest user first to test account linking
        Debug.Log("Requesting new guest identity first");
        this.logOutput.text += "Requesting new identity\n";
        AWSGameSDKClient.Instance.LoginAsNewGuestUser(this.OnGuestLoginResponse);
    }

    // Triggered by the SDK if there's any login error or error refreshing access token
    void OnLoginOrRefreshError(string error)
    {
        Debug.LogError("Login or refresh error: " + error);
        this.logOutput.text += "Login or refresh error: " + error + "\n";

        // NOTE: You would here trigger a new log in or other remediation
    }

    // Update is called once per frame
    void Update()
    {
        
    }

    void OnGuestLoginResponse(LoginRequestData userInfo)
    {
        Debug.Log("New guest Login response: UserID: " + userInfo.user_id + "GuestSecret: " + userInfo.guest_secret + "Token: " + userInfo.auth_token);
        this.logOutput.text += "Login response: \nUserID: " + userInfo.user_id + " \nGuestSecret: " + userInfo.guest_secret + " \n";
        
        // Now let's test linking Facebook ID to the guest user
        // NOTE: You're expected to input a valid Facebook access token and user ID here, see official documentation for Unity Facebook SDK: https://developers.facebook.com/docs/unity/
        //       You will receive a Facebook Access Token and User ID after calling the Login method.
        AWSGameSDKClient.Instance.LinkFacebookIdToCurrentUser("AccessTokenHere","UserIDHere", this.OnLinkFacebookIdResponse);
    }

    void OnLinkFacebookIdResponse(LoginRequestData userInfo)
    {
        Debug.Log("Facebook ID linking response: UserID: " + userInfo.user_id + "FacebookId: " + userInfo.facebook_id + "Token: " + userInfo.auth_token);
        this.logOutput.text += "Facebook ID linking response: \nUserID: " + userInfo.user_id + "\nFacebookId: " + userInfo.facebook_id + " \n";

        // Now let's test logging in with this existing Facebook ID
        // NOTE: As above, you're expected to have a valid Facebook access token and user ID here
        AWSGameSDKClient.Instance.LoginWithFacebookAccessToken("AccessTokenHere", "UserIdHere", this.OnLoginWithFacebookResponse);
    }

    void OnLoginWithFacebookResponse(LoginRequestData userInfo)
    {
        Debug.Log("Facebook ID login response: UserID: " + userInfo.user_id + "FacebookId: " + userInfo.facebook_id + "Token: " + userInfo.auth_token);
        this.logOutput.text += "Facebook ID login response: \nUserID: " + userInfo.user_id + "\nFacebookId: " + userInfo.facebook_id + " \nToken: " + userInfo.auth_token + "\n";
    }
}

File: UnrealSample/Source/UnrealSample/Samples/FacebookLoginTest/FacebookLoginTest.cpp
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0

#include "FacebookLoginTest.h"
#include "../../AWSGameSDK/AWSGameSDK.h"
#include "Engine/World.h"
#include "Kismet/GameplayStatics.h"

// Sets default values for this component's properties
UFacebookLoginTest::UFacebookLoginTest()
{
	// Set this component to be initialized when the game starts, and to be ticked every frame.  You can turn these features
	// off to improve performance if you don't need them.
	PrimaryComponentTick.bCanEverTick = true;

	// ...
}


// Called when the game starts
void UFacebookLoginTest::BeginPlay()
{
	Super::BeginPlay();

	// Get the subsystems
    UGameInstance* GameInstance = Cast<UGameInstance>(UGameplayStatics::GetGameInstance(GetWorld()));
    UAWSGameSDK* AWSGameSDK =  GameInstance->GetSubsystem<UAWSGameSDK>();

    // Init with the login endpoint defined in the Editor and a callback to handle errors for logging in and refresh
	AWSGameSDK->Init(this->m_loginEndpoint);
	AWSGameSDK->OnLoginFailure.AddUObject(this, &UFacebookLoginTest::OnLoginOrRefreshErrorCallback);

	// Define the OnLoginResult callback
	UAWSGameSDK::FLoginComplete loginCallback;
    loginCallback.BindUObject(this, &UFacebookLoginTest::OnGuestLoginResultCallback);

	// Login as a new guest user first
    AWSGameSDK->LoginAsNewGuestUser(loginCallback);
	
}

// Called when there is an error with login or token refresh. You will need to handle logging in again here
void UFacebookLoginTest::OnLoginOrRefreshErrorCallback(const FString& errorMessage){
    UE_LOG(LogTemp, Display, TEXT("Received login error: %s \n"), *errorMessage);
    if(GEngine)
        GEngine->AddOnScreenDebugMessage(-1, 30.0f, FColor::Red, FString::Printf(TEXT("Received login error: \n %s \n"), *errorMessage), false, FVector2D(1.5f,1.5f));

    // NOTE:  You will need to handle logging in again here
}

// Called when guest login is done
void UFacebookLoginTest::OnGuestLoginResultCallback(const UserInfo& userInfo){
    UE_LOG(LogTemp, Display, TEXT("Received guest login response: %s \n"), *userInfo.ToString());
    if(GEngine)
            GEngine->AddOnScreenDebugMessage(-1, 30.0f, FColor::Black, FString::Printf(TEXT("Received login response: \n %s \n"), *userInfo.ToString()), false, FVector2D(1.5f,1.5f));

    // Test linking facebook id to the existing guest account
	UAWSGameSDK::FLoginComplete onLinkFacebookIdCallback;
    onLinkFacebookIdCallback.BindUObject(this, &UFacebookLoginTest::OnLinkFacebookIdResultCallback);
	UGameInstance* GameInstance = Cast<UGameInstance>(UGameplayStatics::GetGameInstance(GetWorld()));
    UAWSGameSDK* AWSGameSDK = GameInstance->GetSubsystem<UAWSGameSDK>();

    // NOTE: You need a valid Facebook access token and user ID that can be obtained by logging in with a Facebook account
    //       We don't currently have any recommendations on how to integrate the Facebook SDK with Unreal Engine 5, but there are some community options available
	AWSGameSDK->LinkFacebookIdToCurrentUser("tokenHere", "userIDHere", onLinkFacebookIdCallback);
}

void UFacebookLoginTest::OnLinkFacebookIdResultCallback(const UserInfo& userInfo){

    UE_LOG(LogTemp, Display, TEXT("Received Facebook ID linking response: %s \n"), *userInfo.ToString());
    if(GEngine)
            GEngine->AddOnScreenDebugMessage(-1, 30.0f, FColor::Black, FString::Printf(TEXT("Received Facebook ID linking response: \n %s \n"), *userInfo.ToString()), false, FVector2D(1.5f,1.5f));

    // Test logging in with existing facebook_id
	UAWSGameSDK::FLoginComplete onFacebookLoginCallback;
    onFacebookLoginCallback.BindUObject(this, &UFacebookLoginTest::OnLoginWithFacebook);
	UGameInstance* GameInstance = Cast<UGameInstance>(UGameplayStatics::GetGameInstance(GetWorld()));
    UAWSGameSDK* AWSGameSDK = GameInstance->GetSubsystem<UAWSGameSDK>();

    // NOTE: As above, you're expected to have a valid Facebook access token and user ID
    AWSGameSDK->LoginWithFacebookAccessToken("tokenHere", "userIdHere", onFacebookLoginCallback);
}

void UFacebookLoginTest::OnLoginWithFacebook(const UserInfo& userInfo){

    UE_LOG(LogTemp, Display, TEXT("Received Facebook login response: %s \n"), *userInfo.ToString());
    if(GEngine)
        GEngine->AddOnScreenDebugMessage(-1, 30.0f, FColor::Black, FString::Printf(TEXT("Received Facebook login response: \n %s \n"), *userInfo.ToString()), false, FVector2D(1.5f,1.5f));
}

// Called every frame
void UFacebookLoginTest::TickComponent(float DeltaTime, ELevelTick TickType, FActorComponentTickFunction* ThisTickFunction)
{
	Super::TickComponent(DeltaTime, TickType, ThisTickFunction);

	// ...
}

